### PR TITLE
Fix some aggregator parse errors

### PR DESCRIPTION
**Pull request by tpoterba**

### CODE REVIEW

File: python/hail/expr/aggregators.py [hunk 1]
```diff
@@ -724,6 +727,8 @@ def filter(condition, expr):
         lambda_result = to_expr(
             condition(
                 construct_expr(Reference(uid), agg._type, agg._indices, agg._aggregations, agg._joins, agg._refs)))
+        if not lambda_result._aggregations.empty():
+            raise ExpressionException("aggregators may not be used inside lambda functions")
```
Comment (danking) [inline python/hail/expr/aggregators.py]: I'm a bit confused by this. Is this OK?

```python
t.aggregate(agg.collect(agg.filter(agg.sum(t.VAL) >= 70, t.ID)
```

it's not really about the lambda

Comment (tpoterba) [reply python/hail/expr/aggregators.py]: that will throw a different error -- can't aggregate an already-aggregated thing. Maybe we don't need this, actually.

File: python/hail/expr/functions.py [hunk 2]
```diff
@@ -303,6 +303,8 @@ def bind(expr, f):
 
     f_input = construct_expr(Reference(uid), expr._type, expr._indices, expr._aggregations, expr._joins, expr._refs)
     lambda_result = to_expr(f(f_input))
+    if not lambda_result._aggregations.empty():
+        raise ExpressionException("aggregators may not be used inside lambda functions")
```
Comment (danking) [inline python/hail/expr/functions.py]: I think the issue is that `f_input` should throw an error if it's placed inside of an aggregation. This should still be OK:

```python
t.annotate(foo = hl.bind(1, lambda x: x + agg.sum(t.VAL)))
```

Maybe names need a "binding context". You can only reference names that are bound _inside_ the aggregator, like this:

```python
t.annotate(foo = agg.mean(hl.bind(... big computationally expensive expression here ...,
                                  lambda x: x * x / t.VAL)))
```

Comment (tpoterba) [reply python/hail/expr/functions.py]: ok, will see if I can rework this.

File: python/hail/expr/aggregators.py [hunk 3]
```diff
@@ -14,17 +14,30 @@ def _to_agg(x):
...
                           aggregations.push(Aggregation(indices, refs)), joins)
 
+def _check_agg_bindings(expr):
+    bound_references = {ast.name for ast in expr._ast.search(lambda ast: isinstance(ast, Reference) and not ast.top_level)}
+    bind_vars = {ast.uid for ast in expr._ast.search(lambda ast: isinstance(ast, Bind))}
+    lambda_vars = {ast.lambda_var for ast in expr._ast.search(lambda ast: isinstance(ast, LambdaClassMethod))}
+    declared_temp_vars = bind_vars.union(lambda_vars)
+
+    referenced_but_not_declared = bound_references - declared_temp_vars
+    if referenced_but_not_declared:
```
[the remark this answers is not in the document]
Comment (tpoterba) [reply python/hail/expr/aggregators.py]: 👍

[the remark this answers is not in the document]
Comment (tpoterba) [reply python/hail/expr/aggregators.py]: fixed this during rebase

**Pull request merged**